### DISCUSSION TITLE
Sectigo SafeNet Authentication Client

### DISCUSSION THREAD

**Discussion by JohnBoyB** [MatthiasValvekens/pyHanko#463] [Q&A] [answered]
Hello, first of all, thank you very much for developing and maintaining this tool.

We have ordered a certificate from Sectigo and received it on a PKCS11 SafeNet USB token. Sectico itself provides a password but no pin. I have set it with the settings 
user-pin: None
prompt-pin: False
Unfortunately I always get the message incorrect pin. I have also tried the DEFER & Skip setting, unfortunately without success. 
Do you have any ideas on how I can use your tool with the password that I enter in the SafeNet Authetication client?

**Comment by JohnBoyB** [comment]
Okay, I can see that probably nobody has similar problems with sectico.
Can anyone tell me about providers where it has worked with a USB stick and PKCS11?

**Comment by aunterste** [comment]
Not yet an issue as we are looking to order a certificate, so adding to this to hear which providers worked.

**Comment by MatthiasValvekens** [comment]
Perhaps a silly suggestion, but did you try supplying the password as the PIN? There's no requirement in the PKCS11 spec that the PIN be a number... :)

**Reply by JohnBoyB** [reply]
Yes I tested it but it didn’t worked.

**Reply by MatthiasValvekens** [reply]
Did you receive any documentation about the PKCS#11 module from the vendor? PKCS#11 is an interface standard, not a hardware specification. In other words, if they're calling this a "PKCS#11 token", what they're actually saying is that they have middleware supporting the PKCS#11 standard API to communicate with the device. If their module doesn't actually implement that specification, you should take it up with them ;).

By the way, which OS are you running this on? If it's Linux: are you sure the issue is not `pinentry`-related? Some PKCS#11 modules are particular about the medium used to enter the PIN/password. Those typically don't accept a PIN supplied through the PKCS#11 API, but will try to spawn  their own dialog through `pinentry`. The defaults are especially likely to cause issues if you're running in a headless environment without a GUI. If this description matches your issue, take a look at the docs for `pinentry` to figure out a configuration that works for you.

**Comment by nautilus7** [comment]
Hi, I recently discovered pyhanko. I am trying to make it work with my safenet 5110 usb token.

I have tried several configurations via the cli, but it always fails. I beleive the problem is the pin.

My token requires a pin only when signing. They call it certificate pin. On windows when I use a pdf editor to sign my files, a window pops up asking for this pin (or rather password).

In my tests, in a Linux server (headless) I haven't managed to feed the token the correct pin to sign the document. I can only supply the token password (what pyhanko calls user pin) but it doesn't like that.

Is there a way to supply the certificate pin via the cli? Is the API able to do that instead?

Thanks.

**Reply by MatthiasValvekens** [reply]
Hi, there are 2 possibilities here. It might be a pinentry plumbing issue (see https://github.com/MatthiasValvekens/pyHanko/discussions/463#discussioncomment-10711353 further up in this discussion) or it could be that the token expects you to re-authenticate on a sign operation with the PKCS#11 API.

The latter is not exposed _conveniently_ within pyHanko (although that could be arranged), but there's an experiment you could try to confirm whether or not that is the problem. Subclass `PKCS11Signer` and override [this method](https://github.com/MatthiasValvekens/pyHanko/blob/master/pkgs/pyhanko/src/pyhanko/sign/pkcs11.py#L542-L550) or monkeypatch it to add a `pin=b"<something>"` parameter to `sign_kwargs` on the returned `PKCS11SignatureOperationSpec` object. Can you confirm whether that works, and whether the signing PIN is the same as the session PIN or not? Depending on those answers I'll work out a way to make that functionality available in the high-level API more conveniently.

As an aside, I'm currently in the process of sourcing hardware from various manufacturers to build a test bench for dealing with this class of vendor-specific issues. This token wasn't on my list yet, but I'll see what I can do about that (my resources are finite, though).

**Reply by nautilus7** [reply]
Hi, sorry for the laye reply. I tried your proposed hack and it worked!

Actually I had to add `sign_kwargs['pin'] = '<something>'` as a unicode str, not a byte str, or else I would get a `AttributeError: 'bytes' object has no attribute 'encode'. Did you mean: 'decode'?` error.

Now, to answer some of your questions:

1. This safenet 5110cc token is used by professionals for signing documents intended for the public sector in the EU (Greece in particular). It is a very popular model. It's a PKCS#11 token.

2. In the particular token I have in hand for testing the user pin (or session pin) and the cert pin (or signing pin) are indeed different.

3. Using a pdf editor or other application for signing I have to enter only the signing pin in a pop app window like these below (either one, depending on the pc I use:

![safenet_token](https://github.com/user-attachments/assets/39cd34a7-aeec-4ebc-adb2-55c85633a684)
<img width="456" height="287" alt="windows_token" src="https://github.com/user-attachments/assets/74dd3850-7452-4980-afe4-5c79530329cf" />

4. Using pyhanko (cli or api) I have to enter both the session pin and the signing pin in order to sign. I was able to provide the session pin using the exposed config/api options. I am not sure, but I think this is a driver thing for security reasons when the token is accessed programmatically. I tried different settings (defer, prompt, etc) without any change.

6. I had no way to enter the signing pin, until I monkey patched that signing method as you proposed. So, it would be great if you could expose this option at a higher level. That would make signing in a headless enviroment possible.

7. I have some questions regarding creating a PAdES-LTA signature and validating it. I am not able to create a signature similar to what adobe acrobat creates. I am unable to understand why, but I think it would be better to create a new discussion and ask there.

8. Finally, I think I came across a bug. I tried validating signatures against EU trusted lists, but the correspong module, does not like reading/saving the trust lists in a path with unicode characters. All reading and writing in this module is done without `encoding='utf-8'` when opening files.

Thanks, pyhanko is great and I look forward for the updates.

**Reply by MatthiasValvekens** [reply]
Hi, thanks for the comments, that clarifies a lot. I've since made some changes (-> `v0.32.0`) that should at least make it possible to provide `sign_kwargs` without having to monkeypatch anything. You can supply them in a JIT fashion by subclassing `PKCS11Signer` and overriding the `sign_kwargs` method, or just pass in `base_sign_kwargs` to `PKCS11Signer` if you want to set them up at initialisation time.


> I have some questions regarding creating a PAdES-LTA signature and validating it. I am not able to create a signature similar to what adobe acrobat creates. I am unable to understand why, but I think it would be better to create a new discussion and ask there.

Yes, please do. PyHanko implements the complete PAdES spec, so I'm sure we can find a solution.

> Finally, I think I came across a bug. I tried validating signatures against EU trusted lists, but the correspong module, does not like reading/saving the trust lists in a path with unicode characters. All reading and writing in this module is done without encoding='utf-8' when opening files.

Judging from your screenshots I think you're on Windows, right? In that case, this isn't a problem with pyHanko, but rather a problem with the default encoding of your operating system, which Python assumes too. You sure it isn't set to something like ISO/IEC 8859-7, since your OS language seems to be set to Greek?

Funny coincidence: [I asked pretty much this exact question over 10 years ago on SO](https://stackoverflow.com/questions/27452317/default-encoding-of-python-3s-open-under-windows-sys-getdefaultencoding-se), but it went unanswered for a long while, and I don't think I ever tried the solution that was posted >6 months later. I decommissioned that Windows machine not long after for unrelated reasons, which is probably why.

**Reply by nautilus7** [reply]
> Yes, please do. PyHanko implements the complete PAdES spec, so I'm sure we can find a solution.

I will thanks.

> Judging from your screenshots I think you're on Windows, right? In that case, this isn't a problem with pyHanko, but rather a problem with the default encoding of your operating system, which Python assumes too. You sure it isn't set to something like ISO/IEC 8859-7, since your OS language seems to be set to Greek?
>
> Funny coincidence: [I asked pretty much this exact question over 10 years ago on SO](https://stackoverflow.com/questions/27452317/default-encoding-of-python-3s-open-under-windows-sys-getdefaultencoding-se), but it went unanswered for a long while, and I don't think I ever tried the solution that was posted >6 months later. I decommissioned that Windows machine not long after for unrelated reasons, which is probably why.

Well, the screenshots were made in windows, but I am testing on various machines including Debian lxc (headless) and windows VM with both English and Greek locales. The problem with the Unicode paths happened on a windows machine with Greek locale. I'll have to check the encoding settings there to confirm whether it fixes the issue.

Thanks again.

**Reply by MatthiasValvekens** [reply]
Alright. In the meantime, I also ordered a SafeNet 5110cc device for my test bench. I hope to have that up and running within the year still--need some more time to design & set up the CI & test orchestration, and there's other hardware I'm still waiting for.

**Reply by nautilus7** [reply]
Great news!

**Reply by MatthiasValvekens** [reply]
Hi @nautilus7, could you specify exactly which of the bundled PKCS#11 modules you use? For obvious reasons I bought a blank token, and it looks like there are several possible setup configurations. I'd like to take a stab at reproducing your setup first.

I tried `libeTPkcs11.so` first, but with that one the key I generated was not marked as sign-exclusive, and the 2nd PIN was equal to the token password. Then I figured it'd have to be the virtual "Digital Signature" slot exposed via `libIDPrimePKCS11.so`, but I'm having some issues getting the keygen to work on that one.

**Reply by nautilus7** [reply]
Hi, not sure I understand exactly what you ask. I'll tell you what I know.

The token I have in hand it's not mine. I haven't initiliazized it personally and I don't know if it came with a preinstalled key or the user created one during initialization. I guess the second should be the case, but I haven't create a key, since it was already initialized by its owner.

I think the best way to initialize the token is with the bundled app ([Safenet authentication client ](https://knowledge.digicert.com/general-information/how-to-download-safenet-authentication-client) - it also contains various pdf guides that might of help) either in linux or in windows. It gives you options to set various passwords. But again, since I haven't done any initialization myself I can't be sure.

I have used `libeTPkcs11.so` for all my testing either in linux or in windows. The behavior regarding the password asking was the same.

Don't know if I answered your questions.

**Reply by MatthiasValvekens** [reply]
That's the thing--the SAC UI allows setting many things, but generating keys isn't one of them :D.

When I generate a key using `libeTPkcs11.so`, curiously enough, I don't see the behaviour you refer to with the different PINs. I do get another PIN prompt when signing, but the pre-signing PIN is the same as the one used to establish the session (and the private key has all sorts of capabilities enabled that wouldn't fly with common criteria requirements, so that's another difference). If I try to generate a key with the IDPrime PKCS#11 module's virtual slot for digital signing, the `C_GenerateKeyPair` call fails.

I'll continue digging.

PS: if you have pkcs11-tool installed, can you show the output of `pkcs11-tool -l -O` for your card? I'm in particular interested in what the `Usage: ` of your private key object indicates.

**Reply by nautilus7** [reply]
hi, the usage is `sign`. This is the complete output:

```
pkcs11-tool --module /usr/lib/libeTPkcs11.so -l -O

Using slot 0 with a present token (0x0)
Logging in to "[redacted]".
Please enter User PIN: 

Certificate Object; type = X.509 cert
  label:      [redacted]
  subject:    DN: C=GR/serialNumber=[redacted], SN=[redacted], GN=[redacted], CN=[redacted]
  serial:     [redacted]
  ID:         [redacted]
  uri:        pkcs11:model=ID%20Prime%20MD;manufacturer=Gemalto;serial=[redacted];token=[redacted];id=%[redacted];object=[redacted];type=cert

Public Key Object; RSA 2048 bits
  label:      [redacted]
  ID:         [redacted]
  Usage:      verify
  Access:     local
  uri:        pkcs11:model=ID%20Prime%20MD;manufacturer=Gemalto;serial=[redacted];token=[redacted];id=%[redacted];object=[redacted];type=public

Private Key Object; RSA 
  label:      [redacted]
  ID:         [redacted]
  Usage:      sign
  Access:     always authenticate, sensitive, always sensitive, never extractable, local
  uri:        pkcs11:model=ID%20Prime%20MD;manufacturer=Gemalto;serial=[redacted];token=[redacted];id=%[redacted];object=[redacted];type=private
```

**Reply by MatthiasValvekens** [reply]
OK, after some trial and error (and some reverse engineering of the PKCS#11 module binary) I found a way to generate a key on-device!

After setting `EnablePrompt=1` in `eToken.conf`, logging into the digital signing slot via the IDPrime PKCS#11 module using the DigSig PIN and calling `C_GenerateKeyPair` with `CKA_SIGN: True` I got a UI prompt for the _regular_ Token PIN. This allowed the request to proceed. So I don't have a way of generating keys in headless mode (yet?), but that's OK, since in principle the setup only needs to be done once. Doesn't help that this procedure isn't documented anywhere, other than (presumably) in internal docs that are shared with CA partners. I know there has to be a way to script this completely, since some CAs ship tools to generate a key + CSR with a SafeNet token during certificate enrolment, but I couldn't piece that together from public documentation.

The good news is that once I had that key set up, I was able to reproduce the behaviour you flagged in headless mode with `pkcs11-tool`: attempting to sign with `libeTPkcs11.so` requires putting in the user PIN and then the DigSig PIN at the context-specific PIN prompt.

But there is actually a way to create signatures with only the DigSig PIN: if you use `libIDPrimePKCS11.so` instead, and log in to the virtual slot labelled `(Digital Signature)`, you can log in to the token & confirm the signature with the DigSig PIN only. This saves you the trouble of having to supply two different PINs at runtime.

**Comment by MatthiasValvekens** [comment, marked as answer]
@nautilus7 @JohnBoyB A few hours ago I added a SafeNet 5110 CC dongle my the HSM test bench, so regular CI builds will be tested against this token too.

As part of the prep work to allow that, I also provided a more "batteries included" way to enable re-authentication on signing operations. `PKCS11SignatureConfig` now has `signing_pin_mode` and `signing_pin` as options. The CLI also supports prompting for the 2nd PIN now, but it currently doesn't attempt to autodetect whether reauthentication is necessary, so you need to explicitly set the `signing_pin_mode` to `PROMPT` for that to work.


The test token is configured in the following manner:

 - Signing key generated on-device, with signing being the only allowed key usage.
 - Signer's certificate + CA chain stored on the token.
 - User token PIN for login + separate digital signature PIN for signing operations.
 - The PKCS#11 module used in testing is `libeToken.so`.
 - I have an RSA 4096 key pair and an ECDSA secp256r1 one configured on the key. Both are used in the test suite.

**Reply by nautilus7** [reply]
Hi, happy new year.

Thank you for the updates and the new possibilities. I will be able to test in a couple of weeks and report back.
